File: MessageTransferTool/src/main/sources/me/xethh/libs/mtt/MessageTransferTool/base/MDCActor.java
package me.xethh.libs.mtt.MessageTransferTool.base;

import akka.actor.AbstractActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import org.slf4j.MDC;
import scala.PartialFunction;
import scala.runtime.BoxedUnit;

import java.util.Map;

public abstract class MDCActor extends AbstractActor {
    protected LoggingAdapter log = Logging.getLogger(getContext().getSystem(), this);

    @Override
    public void aroundReceive(PartialFunction<Object, BoxedUnit> receive, Object msg) {
        if(msg instanceof MDCEntity){
            log.info("Received MDCEntity {}",msg.getClass());
            MDC.clear();
            for(Map.Entry<String,String> entry:((MDCEntity) msg).getMaps().entrySet())
                MDC.put(entry.getKey(),entry.getValue());
            super.aroundReceive(receive, msg);
            MDC.clear();
        }
        else{
            log.info("Received Non MDCEntity {}", msg.getClass());
            super.aroundReceive(receive,msg);
        }
    }
}

File: akka-commons/src/main/sources/me/xethh/libs/toolkits/akka/commons/xeActor/MDCActor.scala
package me.xethh.libs.toolkits.akka.commons.xeActor

import java.util

import me.xethh.libs.toolkits.threadLocal.XEContext
import org.slf4j.MDC

trait MDCActorMessage extends ActorMessage {
  def map():java.util.Map[String,Object]
}
class MDCActorMessageAbs() extends MDCActorMessage {
  val privateMap:java.util.Map[String, Object] = XEContext.copy()
  override def map(): util.Map[String, Object] = privateMap
}
abstract class MDCAroundReceiveActor[AM <: MDCActorMessageAbs] extends AroundReceiveActor[AM] {
  def before(receive:Receive, msg:AM): Unit={
    MDC.clear()
    msg.map().entrySet().forEach{
      entry=> XEContext.put(entry.getKey,entry.getValue)
    }
  }
  def after(receive:Receive, msg:AM): Unit={
    XEContext.copy().entrySet().forEach(x=>XEContext.remove(x.getKey))
  }
}
